New command: "kubeadm token generate"

pipejakob opened the pull request:
As part of #33930, this PR adds a new top-level command to kubeadm to just generate a token for use with the init/join commands. Otherwise, users are left to either figure out how to generate a token on their own, or let `kubeadm init` generate a token, capture and parse the output, and then use that token for `kubeadm join`.

At this point, I was hoping for feedback on the CLI experience, and then I can add tests. I spoke with @mikedanese and he didn't like the original propose of `kubeadm util generate-token`, so here are the runners up:

```
$ kubeadm generate-token          # <--- current implementation
$ kubeadm generate token          # in case kubeadm might generate other things in the future?
$ kubeadm init --generate-token   # possibly as a subcommand of an existing one
```

Currently, the output is simply the token on one line without any padding/formatting:

```
$ kubeadm generate-token
1087fd.722b60cdd39b1a5f
```

CC: @kubernetes/sig-cluster-lifecycle 

**Release note**:

<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->

``` release-note
New kubeadm command: generate-token
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35144)

<!-- Reviewable:end -->

dgoodwin reviewed (commented):
Not 100% sure where https://github.com/kubernetes/kubernetes/pull/30707 will lang yet, but I suspect kubeadm will be the means by which we create additional tokens, list the ones in play, and revoke them. 

With that in mind my suggestion would be a full "token" sub-command:

kubeadm token generate
kubeadm token create (to send to server, with optional --token if you pre-created a token to send)
kubeadm token list
kubeadm token revoke [ID]

pires commented:
I'm against adding more features without tests. Can we make this a rule, @mikedanese @jbeda @luxas?

That said, I agree with @dgoodwin opinion on `kubeadm token` command.

pipejakob commented:
@dgoodwin Great suggestion, that seems a lot cleaner. I'll try to get up to speed on that proposal.

@pires Yup, I was just trying to get early feedback on the CLI experience at this point, since any changes there would require refactoring the code (and hence the tests). Let me know if there's a better way to signal "I'm looking for early feedback, but I know this commit isn't actually ready to be merged." I'll try to make that even more explicit in the future. And don't worry, we're on the same page: tests are good. Kubeadm could use a lot more of them, and I'll try to do my part :-).

pires commented:
Thanks a ton @pipejakob!

pires commented:
Ah, you can name your Pr as WIP or DON'T MERGE. Then a maintainer will mark it so the bot doesn't merge it.

jbeda commented:
I like the plan of having `kubeadm token *` commands.  `kubeadm token generate` seems like a great place to start with that.  Also agree we'll want to have tests as this goes in.

As part of the documentation for `kubeadm token generate --help` we should make it clear that this is just a random string that of the form `<6chars>.<16 chars>` that can be generated easily in other ways and that this command is just a convenience.  I don't want folks to think that there is anything super special about how kubeadm is doing it.

dgoodwin commented:
@pipejakob I'm going to start work on some code to create and manage the token secrets per plans in #30707, just wanted to let you know and I will keep an eye on this PR and either build on top of your commits, or in a parallel structure so the two can be easily merged later.

pipejakob commented:
I updated the pull request:

* The command is now "kubeadm token generate," as suggested.
* I've added unit and CLI-level tests for the new command.
* In order to find the kubeadm binary for testing, I copied the logic from `kubectl.sh` to the new `kubeadm.sh`. I broke up the PR into several commits so you can follow exactly what I did and what has changed.
* I deliberately did not add new CLI tests to the existing `test-cmd.sh` file because 1) it is already extremely long, 2) I wanted to use Go for testing instead of bash, and 3) to prepare for eventually pulling kubeadm out of this repo entirely per #35314.

Barring code review feedback, I think this is now actually ready to be merged, so can someone remove the do-not-merge label for me? Thank you!

dmmcquay commented on cmd/kubeadm/app/cmd/token_cmd_test.go:
```diff
@@ -0,0 +1,67 @@
+// +build cmd
```
This is all really cool! Really glad to see this and more testing going into place.  I would just suggest that we shouldn't have a `token_cmd_test.go` file, since that doesn't follow the Go standard for naming test files.  I think everything inside of `token_cmd_test.go` should just get placed in `token_test.go`.

pipejakob replied:
Good point. That does mean I won't be able to use the +build directive to separate the tests, but I think I can get by with just using a specific naming scheme for the tests and doing "go test -run Cmd" instead of "go test -tags cmd". I should have a new commit soon.

dgoodwin commented:
I have additional token sub-commands cooking in my branch, will merge them in with this code when it merges.

luxas commented:
OK, @jbeda @dgoodwin @pires can you take another look?

luxas commented on cmd/kubeadm/app/cmd/token.go:
```diff
@@ -0,0 +1,86 @@
+}
+
+func RunGenerateToken(out io.Writer) error {
+	s := new(kubeadmapi.Secrets)
```
nit: `kubeadmapi.Secrets{}`

luxas commented on cmd/kubeadm/app/cmd/token_test.go:
```diff
@@ -0,0 +1,102 @@
+var kubeadmPath string
+
+func init() {
+	flag.StringVar(&kubeadmPath, "kubeadm-path", "", "Location of kubeadm")
```
This seems much more like integration tests than unit tests to me.

Do everyone think that we should have integration tests yet?
Just wondering, because they probably should be in another package totally, and share this logic.

Because we don't want to have every `_test.go` file to have this `kubeadm-path` logic in it, and it doesn't really feel ideomatic what we do below:
```
make -C "${KUBE_ROOT}" test \
  WHAT=cmd/kubeadm/app/cmd \
  KUBE_TEST_ARGS="-run Cmd --kubeadm-path '${KUBEADM_PATH}'"
```

It would be much better if you could split this logic into a separate pkg; initially maybe `cmd/kubeadm/integration` or something.

And I expect unit tests (i.e. `token_test.go`, `init_test.go`, etc.) to call the functions of the main file directly (i.e. `token.go` and `init.go`) like this:
```go
func TestRunGenerateToken (t *testing.T) {
   // init mywriter
   err := RunGenerateToken(mywriter)
   // parse output written to mywriter
   // etc...
}
```

pires replied:
These are not integration tests but rather functional tests. I vote to follow the same approach as `kubectl`.

jbeda replied:
I'd encourage us to keep `kubeadm` isolated as much as possible from centralized test infra.  We will eventually want to move kubeadm out to its own repo (after kubectl moves out).

luxas replied:
What I'm saying is that we should move the tests that are executing a pre-built kubeadm binary into `cmd/kubeadm/test/integration` or a similar path (`cmd/kubeadm/app/integration`)

And in those tests we can execute the binary and verify that functionality is working properly.

In "normal" `gofile_test.go` files, we should have unit tests that call functions in `gofile.go`, just like @dgoodwin has done in https://github.com/kubernetes/kubernetes/pull/35632

@dmmcquay No we shouldn't centralize these integration test to `test/integration`, but have  kubeadm-only tests that are easy to take with us if/when we move to a separate repo

@pipejakob Do you know what I mean with this

dmmcquay replied:
@luxas if you take a look at my comment below, I think it explains pretty well what I think should happen. I am not suggesting that we centralize them, just showing what had previously been done. Also, I don't think these are integration tests and I think the correct way to call out to subprocess and verify things I've linked below:

https://play.golang.org/p/Qg_c1eRKur

pipejakob replied:
TL;DR: I'd like to continue categorizing these tests as `cmd` tests, have them run during the `make test-cmd` target, and create a new tree for them to live in, `kubernetes/test/cmd`.

So, there's a lot of discussion around exactly what kind of tests these are and where they should live. I think there are a few dimensions at play:

- What to call them
- When they should execute
- Where the code should live

To my understanding, we currently have four kinds of tests, with different Makefile targets for executing each kind:

| Target | Description |
| --- | --- |
| make test | Unit tests |
| make test-integration | Integration tests |
| make test-cmd | CLI tests |
| make test-e2e | End-to-end tests |

(From [build-tools/README.md](https://github.com/kubernetes/kubernetes/blob/e09fc6d5b91a28f4beecd6c73e2c44c2340d3bd3/build-tools/README.md) and [Makefile](https://github.com/kubernetes/kubernetes/blob/e09fc6d5b91a28f4beecd6c73e2c44c2340d3bd3/Makefile#L116))

In spirit, these tests fit in very nicely with the existing `test-cmd` CLI tests that exist to test `kubectl`, `kubelet`, `kube-apiserver`, etc. Like those tests, mine require that the binaries be built before the tests can run, and they actually test the functionality at the CLI level to catch things that unit tests might miss (or can't test), just like the existing `test-cmd` tests.

The only reason I didn't add them to the existing `test-cmd` tests is because 1) all of those tests are in an enormous monolithic file that makes it difficult to selectively run subsets of tests, 2) the file is all bash, and 3) it seemed counter-productive to pulling `kubeadm` out to its own repo in the future. (Here is that file for reference: [test-cmd.sh](https://github.com/kubernetes/kubernetes/blob/e09fc6d5b91a28f4beecd6c73e2c44c2340d3bd3/hack/make-rules/test-cmd.sh))

If it weren't for those things, I would have just added these tests directly to `test-cmd.sh` and I don't think there would have been much to discuss. So to answer the question of "what kinds of tests they are," I would call them `cmd` tests. I don't think anyone is arguing that they're unit tests, and they're far from what we call end-to-end tests (which actually spin up an entire cluster for testing). It's debatable if they should be called integration tests, but our existing integration tests don't test at the CLI level, and moreover, these tests aren't really testing the integration of multiple components, they're just testing the same simple component at a higher level. We may absolutely have kubeadm integration tests in the future, but I don't think these really qualify in spirit or match the expectations of our existing `make test-integration` target. So, I would like to continue calling them `cmd` tests and having them execute as part of `make test-cmd`.

As for where the tests should actually live in the codebase, I originally had them in a separate test file in the `cmd/kubeadm/app` tree, alongside the unit test file, and only merged them into the unit test file upon request. I do, however, think that there's merit to separating out the different kinds of tests we have at the file level, as all of our existing test types do (unit test files are stored right alongside the implementing code, integration tests have their own tree, and `cmd` and `e2e` tests are completely separate in the `hack` tree). That makes it easy to find all of the unit tests in the same place, and understand how to add a unit test by following the same pattern, and the same goes for integration tests, etc. If we merge multiple different kinds of tests into the same file, it makes it easier to shoot yourself in the foot and either have your test run as part of the wrong make test target, have it run redundantly in multiple make test targets, or even worse, accidentally be run in none of our make test targets and silently get ignored because you followed the wrong convention and didn't realize it never ran. I think having different kinds of tests in the same file is just asking for trouble.

So, what I'd propose is to create a new tree in `test/` akin to the existing `test/integration` and `test/e2e` trees: `test/cmd`. This keeps it separate from the other test types and makes it very easy to find all of the `cmd` tests to run during the `make test-cmd` target. I assume the only reason this doesn't exist right now is because of the fact that all of the existing `cmd` tests were written in a single bash file, so there was no need for further organization or hierarchy.

luxas reviewed (commented):
So to recap, I think you should refactor out what you have now in `token_test.go` to an integration package and then we might add other behavioral tests there later.

And in this PR just add unit tests for `token.go` that verifies that the `RunGenerateToken` function works.

I appreciate your work on this. Thanks

dmmcquay commented:
So, turns out that I wasn't understanding exactly what @pipejakob was doing.  I think its a really good idea and I have some suggestions.

https://play.golang.org/p/Qg_c1eRKur

I am not suggestion that the code I wrote is what we should go with, but it at least gives an idea of what I would suggest.  I think the work @pipejakob did with the `*.sh` files is great and I think makes testing the CLI aspect of kubeadm great. What I added in is based on hashicorp's advanced testing with go talk (https://speakerdeck.com/mitchellh/advanced-testing-with-go).  Having a way to check if the actual binary exists (whether we go with checking if it exists in the path or if it exists in the -kubeadm-path flag) and then either running the functional tests or skipping them.

The reason for the really long testing times I mentioned above were actually because the e2e framework pkg that was being imported is absolutely huge.  It does way more than we actually need, and in the example go playground above, I demonstrated that just taking three functions from that pkg would get the same behavior without the behemoth testing times.  

How we decide to handle tests where we don't have the kubeadm binary in place is up for debate.  I personally don't mine just skipping those tests (as the go playground code shows).  I am currently working on adding testing to a lot of the other kubeadm packages and as I am doing so, I typically just use the go tool to test the code than the makefile provided in the root of the code base, and if I don't happen to have a built binary of kubeadm in my path or don't provide the path, I would prefer just to skip those for quick iteration.  Obviously, this would be caught and tested when using the makefile. 

Also, having it check for the binary and decide whether or not to test based on that, we can hopefully get rid of the need to add the `-run cmd` test handler regex when we want to test CLI functionality for kubeadm. 

Hopefully that makes sense and I would be happy to dive deeper into any issues.

/cc @pires @luxas @dgoodwin @jbeda

jbeda commented on cmd/kubeadm/app/cmd/token.go:
```diff
@@ -0,0 +1,86 @@
+		// cobra will print usage information, but still exit cleanly.
+		// We want to return an error code in these cases so that the
+		// user knows that their command was invalid.
+		RunE: func(cmd *cobra.Command, args []string) error {
```
I'm not sure this is worth it.  Other places where a sub-command isn't given we just print help and exit with 0.  (for example `kubeadm`).  In the case with a bad subcommand -- cobra will errror and give suggestions if you have a typo.

I'd just leave RunE undefined.

jbeda commented on hack/make-rules/test-kubeadm-cmd.sh:
```diff
@@ -0,0 +1,31 @@
+
+KUBE_ROOT=$(dirname "${BASH_SOURCE}")/../..
+
+KUBEADM_PATH="${KUBEADM_PATH:=$(realpath "${KUBE_ROOT}")/cluster/kubeadm.sh}"
```
realpath isn't available on non-GNU based systems (mac OS X) and so we shouldn't assume it here.  `hack/lib/init.sh` has a helper that emulates it.

pipejakob replied:
Ah, good catch! I'll fix that.

pipejakob commented:
@k8s-bot verify test this

jbeda commented:
You need to run `hack/update-bazel.sh`

luxas commented on test/cmd/kubeadm/token_test.go:
```diff
@@ -0,0 +1,65 @@
+limitations under the License.
+*/
+
+package kubeadm
```
Thanks for doing this. Great!

However, I think it would be better if this was under `cmd/kubeadm/test` instead, can you do that?
In that case, we have everything centralized under `cmd/kubeadm` where we have the right OWNERS/assignees etc.

pipejakob replied:
Sure!

luxas commented on cmd/kubeadm/app/cmd/token.go:
```diff
@@ -0,0 +1,86 @@
+		`),
+		Run: func(cmd *cobra.Command, args []string) {
+			err := RunGenerateToken(out)
+			cmdutil.CheckErr(err)
```
Se other commands' `Run` function, we're using `kubeadmutil.CheckErr` now

pipejakob commented:
Yeah, I'm working on getting `update-bazel.sh` to work now. When I first ran it locally, it showed no problems (because I haven't pulled from upstream in a few days). After pulling, it's now spewing dependency errors that I'm working through. Thanks! I'll ping @mikedanese for help, since it seems to be coming from his new gazel tool.

luxas reviewed (commented):
I like this patch.

Thanks for addressing most of the earlier comments
Still I have two nits that would be nice to get addressed, before merging.

Also, please squash your commits logically, e.g. the `kubeadm token` go code with unit tests first, and then the `cmd/kubeadm/test` code and lastly the bash code.
Two or three commits should be fine

pipejakob committed:
New command: "kubeadm token generate"

This surfaces the token generation logic so that users can first
generate and store a token, then pass it to kubeadm init/join.
Otherwise, users have to capture and parse the output of "kubeadm init"
to feed the token to "kubeadm join."

pipejakob committed:
Cloning kubectl.sh logic into new kubeadm.sh.

pipejakob committed:
Adding CLI tests for kubeadm.

pipejakob commented:
Hey @luxas, I think I addressed your latest comments. Can you take another look?

luxas commented:
This looks good to me, but I'd like to get the next stable release out of the door first (probably tomorrow)

See: https://github.com/kubernetes/kubernetes/issues/34884

A minor thing, I think that `kubeadm token create` is better as @dgoodwin has in https://github.com/kubernetes/kubernetes/pull/35805, but I think @dgoodwin can/should just rename that in his PR

pipejakob commented:
@k8s-bot node e2e test this

pipejakob commented:
@luxas Got it. I'll leave it as "token generate" and let @dgoodwin change if necessary in his own PR. Thanks again for all the feedback!

dgoodwin commented:
Sounds good to me, I will sort it out, I think there's room for two commands, one to just print a valid token and one to submit to server, but maybe we can combine it into kubeadm token create --print or something like that. (ideas welcome)

pipejakob commented:
@k8s-bot node e2e test this

pipejakob commented:
@k8s-bot node e2e test this

luxas commented:
@k8s-bot unit test this

luxas commented:
LGTM, will merge when we've gotten the release out of the door
